### PR TITLE
:seedling: WaitForDefaultNamespace while starting up envtest

### LINKED ISSUE #2626: Seeing flakey testing where `default` namespace doesn't exist
Currently, we are running `envtest` in Karpenter where we are starting up all of the binaries and then applying pods to the `default` namespace. Every now and then, we are seeing failures in our pod metric testing (seen here: https://github.com/kubernetes-sigs/karpenter/actions/runs/7251475446/job/19753909305?pr=885). When this happens, it always contains the same error

```bash
Pod Metrics [It] should update the pod state metrics
/home/runner/work/karpenter/karpenter/pkg/controllers/metrics/pod/suite_test.go:56

  [FAILED] Expected success, but got an error:
      <*errors.StatusError | 0xc000134aa0>: 
      namespaces "default" not found
      {
          ErrStatus: {
              TypeMeta: {Kind: "", APIVersion: ""},
              ListMeta: {
                  SelfLink: "",
                  ResourceVersion: "",
                  Continue: "",
                  RemainingItemCount: nil,
              },
              Status: "Failure",
              Message: "namespaces \"default\" not found",
              Reason: "NotFound",
              Details: {Name: "default", Group: "", Kind: "namespaces", UID: "", Causes: nil, RetryAfterSeconds: 0},
              Code: 404,
          },
      }
  In [It] at: /home/runner/work/karpenter/karpenter/pkg/controllers/metrics/pod/suite_test.go:58 @ 12/18/23 17:16:52.[71](https://github.com/kubernetes-sigs/karpenter/actions/runs/7251475446/job/19753909305?pr=885#step:5:72)9
```

This error indicates that the `default` namespace doesn't exist at the apiserver when we are applying the object; however, I would generally expect this namespace to exist once the apiserver starts up.

It's worth noting that we don't see this error in other testing where we are running `envtest.Environment` with the `CRDs` option set. I suspect that this is due to the fact that this gives the setup a little more time to populate this namespace at the apiserver before it returns. 

Is there a general recommendation here? Obviously we could wait for the `default` namespace to exist in our own code, but it would be nice if we could be assured that it would always exist when we start testing from upstream.

### PR DESCRIPTION
**Pull request by jonathan-innis**
Fixes #2626

This change adds a `waitForDefaultNamespace` call which polls the apiserver for the existence of the default namespace immediately after control plane startup. This is needed because we've seen that there can be flakes when leveraging `envtest` if you expect the `default` namespace to be there immediately after startup but the namespace hasn't actually propagated so that the apiserver can see it.

### CODE REVIEW

File: pkg/envtest/server.go [hunk 1]
```diff
@@ -322,6 +333,20 @@ func (te *Environment) startControlPlane() error {
 	return nil
 }
 
+func (te *Environment) waitForDefaultNamespace(config *rest.Config) error {
+	cs, err := client.New(config, client.Options{})
+	if err != nil {
+		return fmt.Errorf("unable to create client: %w", err)
+	}
+	// It shouldn't take longer than 5s for the default namespace to be brought up in etcd
+	return wait.PollUntilContextTimeout(context.TODO(), time.Millisecond*500, time.Second*5, true, func(ctx context.Context) (bool, error) {
```
Comment (sbueringer) [inline pkg/envtest/server.go:342]:
```suggestion
	return wait.PollUntilContextTimeout(context.TODO(), time.Millisecond*50, time.Second*5, true, func(ctx context.Context) (bool, error) {
```

Maybe let's do this a bit quicker so we don't lose a full 500ms if the namespace doesn't show up immediately

Comment (jonathan-innis) [reply pkg/envtest/server.go]: Yep, makes sense. Updated.

File: pkg/envtest/server.go [hunk 2]
```diff
@@ -334,12 +334,13 @@ func (te *Environment) startControlPlane() error {
 }
 
 func (te *Environment) waitForDefaultNamespace(config *rest.Config) error {
+	fmt.Println("waiting...")
```
Comment (sbueringer) [inline pkg/envtest/server.go:337]: If we want to add a log here. Let's please also write what we are waiting for and log the same way as in other places (e.g. l.299: `log.V(1).Info("installing webhooks")`)

Comment (jonathan-innis) [reply pkg/envtest/server.go:337]: Shoot. Yep, you're totally right. I don't even think that we want to log here. This was just an ad-how change that I forgot to remove before pushing

Comment (jonathan-innis) [reply pkg/envtest/server.go]: Updated.

**Comment by sbueringer** [comment]
Thx!

/lgtm
/approve

**Pull request merged**